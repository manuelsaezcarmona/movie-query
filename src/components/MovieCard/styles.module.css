.movie-card {
  display: flex;
  flex-direction: column;
  transition: 300ms;
  background: #ffffff;
  box-shadow: var(--shadow-card);
  width: 100%;
  max-width: 315px;

  margin: 2em;
  border-radius: 10px;
}
.movie-card:hover {
  transform: scale(1.03);
  box-shadow: 0px 10px 25px rgba(0, 0, 0, 0.08);
}

.movie-card__body {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 15vh;
  row-gap: 0.5em;
  padding: 1em;
}

.movie-card__image {
  width: 100%;
  height: 40vh;
  object-fit: cover;
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
}

.movie-card__title {
  font-size: 24px;
  font-weight: 400;
  margin: 0;
}

.movie-card__data {
  display: flex;
  justify-content: space-between;
  text-transform: uppercase;
}

.movie-card__datelabel,
.movie-card__ratelabel {
  color: rgba(0, 0, 0, 0.5);
  margin-bottom: 0.5em;
  font-size: 9px;
}

.movie-card__date,
.movie-card__rate {
  font-weight: 700;
  font-size: 11px;
}

.movie-card__rate {
  text-align: center;
}

.bookmark {
  font-size: 1.5em;
}

@media (min-width: 769px) {
  .movie-card {
    width: 80%;
    max-width: 315px;
  }
}
