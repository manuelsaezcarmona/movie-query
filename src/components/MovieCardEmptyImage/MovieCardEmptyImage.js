export default function MovieCardEmptyImage() {
  return (
    <svg xmlns="http://www.w3.org/2000/svg" width={64} height={64}>
      <path d="M61.17 63.509c-.54 0-1.08-.207-1.492-.622L1.338 4.07a2.1 2.1 0 1 1 2.984-2.958l58.34 58.817a2.1 2.1 0 0 1-1.492 3.58zM50.43 21.304c0-1.231.99-2.23 2.211-2.23 1.222 0 2.212.999 2.212 2.23 0 1.232-.99 2.23-2.212 2.23a2.221 2.221 0 0 1-2.212-2.23" />
      <path d="M52.402 54.669H8.447c-4.386 0-7.956-3.59-7.956-8.005V19.575c0-4.413 3.57-8.005 7.956-8.005h3.188a3.29 3.29 0 0 0 1.132-.2 2.1 2.1 0 0 1 1.443 3.946 7.468 7.468 0 0 1-2.575.455H8.447c-2.07 0-3.755 1.707-3.755 3.804v27.09c0 2.096 1.685 3.803 3.755 3.803h43.954a2.1 2.1 0 0 1 0 4.201z" />
      <path d="M32 47.442c-8.455 0-15.334-6.928-15.334-15.442 0-4.121 1.593-7.997 4.486-10.913a2.1 2.1 0 1 1 2.983 2.959A11.22 11.22 0 0 0 20.868 32c0 6.199 4.993 11.241 11.132 11.241 2.97 0 5.763-1.167 7.865-3.286a2.1 2.1 0 1 1 2.983 2.958c-2.897 2.92-6.749 4.53-10.848 4.53zM60.296 52.225a2.101 2.101 0 0 1-1.704-3.326 3.809 3.809 0 0 0 .716-2.235V19.575c0-2.097-1.685-3.804-3.755-3.804h-3.188a7.515 7.515 0 0 1-6.649-4.013l-.963-1.814A4.699 4.699 0 0 0 40.6 7.429H23.401a4.688 4.688 0 0 0-3.568 1.651 2.1 2.1 0 1 1-3.195-2.727 8.882 8.882 0 0 1 6.764-3.126H40.6a8.89 8.89 0 0 1 7.865 4.75l.963 1.813a3.323 3.323 0 0 0 2.937 1.78h3.187c4.386 0 7.956 3.592 7.956 8.005v27.09c0 1.694-.52 3.313-1.504 4.684a2.099 2.099 0 0 1-1.709.876z" />
      <path d="M44.992 36.661a2.102 2.102 0 0 1-2.065-2.498c.137-.71.206-1.438.206-2.164 0-6.198-4.994-11.24-11.132-11.24-.75 0-1.499.074-2.224.223a2.1 2.1 0 1 1-.843-4.115 15.328 15.328 0 0 1 3.067-.31c8.455 0 15.333 6.927 15.333 15.442 0 .992-.094 1.987-.28 2.957a2.104 2.104 0 0 1-2.062 1.705z" />
    </svg>
  );
}
