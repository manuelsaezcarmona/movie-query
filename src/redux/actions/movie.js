import { getMoviesByTerm } from '../../services/http';
import { types } from '../actiontypes';

export const getAllMovies = (result) => ({
  type: types.MOVIE_GET_ALLMOVIES,
  payload: result
});

export const startGetAllMovies = (term, page) => async (dispatch) => {
  try {
    const result = await getMoviesByTerm(term, page);

    if (!result.success) {
      return result.status_message;
    }
    dispatch(getAllMovies(result));
    return result;
  } catch (error) {
    return error;
  }
};
