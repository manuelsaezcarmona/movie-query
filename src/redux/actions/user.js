import { types } from '../actiontypes';

export const addFavourite = (movie) => ({
  type: types.USER_ADD_FAVOURITES,
  payload: movie
});

export const deleteFavourite = (movieID) => ({
  type: types.USER_REMOVE_FAVOURITES,
  payload: movieID
});
