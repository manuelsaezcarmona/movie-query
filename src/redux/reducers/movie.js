import { types } from '../actiontypes';

const initialState = {
  page: 0,
  movies: [],
  activemovie: {},
  totalpages: 0,
  detailedmovie: {}
};

export const movieReducer = (state = initialState, action) => {
  switch (action.type) {
    case types.MOVIE_GET_ALLMOVIES:
      return {
        ...state,
        page: action.payload.page,
        movies: action.payload.results,
        totalpages: action.payload.total_pages
      };
    case types.MOVIE_SET_ACTIVE_MOVIE:
      return {
        ...state,
        activemovie: action.payload
      };
    case types.MOVIE_SET_DETAILED_MOVIE: {
      // console.log(action.payload);
      return {
        ...state,
        detailedmovie: action.payload
      };
    }
    default:
      return state;
  }
};
