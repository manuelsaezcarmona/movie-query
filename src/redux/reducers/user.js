import { types } from '../actiontypes';

const initialState = {
  username: 'Manu',
  favourites: []
};

export const userReducer = (state = initialState, action) => {
  switch (action.type) {
    case types.USER_ADD_FAVOURITES:
      return {
        ...state,
        favourites: [...state.favourites, action.payload]
      };

    default:
      return state;
  }
};
