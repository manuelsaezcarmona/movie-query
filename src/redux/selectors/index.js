import { useSelector } from 'react-redux';

export const selectFavMovies = () => useSelector((state) => state.user.favourites);

export const selectFavMovieById = (id) => {
  const favs = useSelector((state) => state.user.favourites);
  return !!favs.find((fav) => fav.id === id);
};
