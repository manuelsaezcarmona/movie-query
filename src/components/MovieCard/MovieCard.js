import React from 'react';
import { BsBookmarkPlus, BsBookmarkDashFill } from 'react-icons/bs';
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { DEFAULT_WIDTH_POSTER, IMAGE_URL } from '../../services/constants';
import styles from './styles.module.css';
import MovieCardEmptyImage from '../MovieCardEmptyImage/MovieCardEmptyImage';
import { setActiveMovie } from '../../redux/actions/movie';
import { addFavourite, deleteFavourite } from '../../redux/actions/user';
import { selectFavMovieById } from '../../redux/selectors/index';

export default function MovieCard({ movie }) {
  const imgPosterPath = `${IMAGE_URL}/${DEFAULT_WIDTH_POSTER}${movie.poster_path}`;

  const dispatch = useDispatch();

  const handleAddFavourite = () => {
    dispatch(setActiveMovie(movie));
    dispatch(addFavourite(movie));
  };

  const handleDeleteFavourite = () => {
    dispatch(setActiveMovie(movie));
    dispatch(deleteFavourite(movie.id));
  };

  const favMovie = selectFavMovieById(movie.id);

  return (
    <article className={styles['movie-card']}>
      <header className="movie-card__header">
        <Link to={`/id/${movie.id}`}>
          {movie.poster_path ? (
            <img
              className={styles['movie-card__image']}
              src={imgPosterPath}
              alt="Jack el destripador"
            />
          ) : (
            <MovieCardEmptyImage className={styles['movie-card__image']} />
          )}
        </Link>
      </header>

      <div className={styles['movie-card__body']}>
        <h4 className={styles['movie-card__title']}>{movie.title}</h4>
        <div className={styles['movie-card__data']}>
          <div className={styles['movie-card__datecontainer']}>
            <p className={styles['movie-card__datelabel']}>Estreno</p>
            <p className={styles['movie-card__date']}>{movie.release_date}</p>
          </div>

          {!favMovie ? (
            <BsBookmarkPlus className={styles.bookmark} onClick={handleAddFavourite} />
          ) : (
            <BsBookmarkDashFill className={styles.bookmark} onClick={handleDeleteFavourite} />
          )}

          <div className={styles['movie-card__ratecontainer']}>
            <p className={styles['movie-card__ratelabel']}>Valoraci√≥n</p>
            <p className={styles['movie-card__rate']}>{movie.vote_average}</p>
          </div>
        </div>
      </div>
    </article>
  );
}
