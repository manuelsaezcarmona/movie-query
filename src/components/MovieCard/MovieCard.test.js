/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { Provider, useDispatch } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { Link, MemoryRouter } from 'react-router-dom';
import { fireEvent, render, screen } from '@testing-library/react';

import { BsBookmarkPlus, BsBookmarkDashFill } from 'react-icons/bs';

import { DEFAULT_WIDTH_POSTER, IMAGE_URL } from '../../services/constants';
import MovieCardEmptyImage from '../MovieCardEmptyImage/MovieCardEmptyImage';
import { setActiveMovie } from '../../redux/actions/movie';
import { addFavourite, deleteFavourite } from '../../redux/actions/user';
import { selectFavMovieById } from '../../redux/selectors/index';
import { movieDetail } from '../../services/mockdata';

import MovieCard from './MovieCard';

jest.mock('../../redux/actions/movie', () => ({
  setActiveMovie: jest.fn()
}));

jest.mock('../../redux/actions/user', () => ({
  addFavourite: jest.fn(),
  deleteFavourite: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {
  user: {
    username: 'test',
    favourites: []
  },
  movie: {
    page: 1,
    movies: [],
    activemovie: {},
    totalpages: 3,
    detailedmovie: {}
  },
  settings: {
    currentpage: 1,
    searchfilm: 'Avengers'
  }
};

const movieObj = {
  movie: {
    ...movieDetail
  }
};

const store = mockStore(initialState);
store.dispatch = jest.fn();

describe('Given the MovieCard component', () => {
  test('The component is rendered (take a snapshot)', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter>
          <MovieCard {...movieObj} />
        </MemoryRouter>
      </Provider>
    );
    expect(asFragment(<MovieCard />)).toMatchSnapshot();
  });
  test('When add Favourite dispatch', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <MovieCard {...movieObj} />
        </MemoryRouter>
      </Provider>
    );
    const iconAddFavourite = screen.getByTitle('bookmark-plus');
    fireEvent.click(iconAddFavourite);
    expect(setActiveMovie).toHaveBeenCalled();
  });
});
