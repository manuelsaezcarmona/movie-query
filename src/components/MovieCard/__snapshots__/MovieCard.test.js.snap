// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the MovieCard component The component is rendered (take a snapshot) 1`] = `
<DocumentFragment>
  <article
    class="movie-card"
  >
    <header
      class="movie-card__header"
    >
      <a
        href="/id/343611"
      >
        <img
          alt="Jack el destripador"
          class="movie-card__image"
          src="https://image.tmdb.org/t/p/w342/wxLUQ1pIms3HAlVGYvEG9zg2kDs.jpg"
        />
      </a>
    </header>
    <div
      class="movie-card__body"
    >
      <h4
        class="movie-card__title"
      >
        Jack Reacher: Never Go Back
      </h4>
      <div
        class="movie-card__data"
      >
        <div
          class="movie-card__datecontainer"
        >
          <p
            class="movie-card__datelabel"
          >
            Estreno
          </p>
          <p
            class="movie-card__date"
          >
            2016-10-19
          </p>
        </div>
        <svg
          class="bookmark"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            bookmark-plus
          </title>
          <path
            d="M2 2a2 2 0 0 1 2-2h8a2 2 0 0 1 2 2v13.5a.5.5 0 0 1-.777.416L8 13.101l-5.223 2.815A.5.5 0 0 1 2 15.5V2zm2-1a1 1 0 0 0-1 1v12.566l4.723-2.482a.5.5 0 0 1 .554 0L13 14.566V2a1 1 0 0 0-1-1H4z"
          />
          <path
            d="M8 4a.5.5 0 0 1 .5.5V6H10a.5.5 0 0 1 0 1H8.5v1.5a.5.5 0 0 1-1 0V7H6a.5.5 0 0 1 0-1h1.5V4.5A.5.5 0 0 1 8 4z"
          />
        </svg>
        <div
          class="movie-card__ratecontainer"
        >
          <p
            class="movie-card__ratelabel"
          >
            Valoraci√≥n
          </p>
          <p
            class="movie-card__rate"
          >
            5.9
          </p>
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
